app/controllers/ileQuery/IleQueryController.scala

```diff
@@ -0,0 +1,140 @@
+import javax.inject.{Inject, Singleton}
+import models.UcrBlock
+import models.cache.{Answers, IleQuery}
+import models.notifications.queries.IleQueryResponseExchangeData.{SuccessfulResponseExchangeData, UcrNotFoundResponseExchangeData}
+import models.notifications.queries.IleQueryResponseExchange
+import play.api.Logger
+import play.api.i18n.I18nSupport
+import play.api.libs.json.Json
+import play.api.mvc.{Action, AnyContent, MessagesControllerComponents, Request, Result}
+import repositories.IleQueryRepository
+import uk.gov.hmrc.http.HeaderCarrier
+import uk.gov.hmrc.play.bootstrap.controller.FrontendController
+import utils.FieldValidator.validDucr
+import views.html._
+
+import scala.concurrent.{ExecutionContext, Future}
+
+@Singleton
+class IleQueryController @Inject()(
+  authenticate: AuthenticatedAction,
+  mcc: MessagesControllerComponents,
+  errorHandler: ErrorHandler,
+  ileQueryRepository: IleQueryRepository,
+  connector: CustomsDeclareExportsMovementsConnector,
+  ileQueryPage: ile_query,
+  loadingScreenPage: loading_screen,
+  ileQueryDucrResponsePage: ile_query_ducr_response,
+  ileQueryMucrResponsePage: ile_query_mucr_response
+)(implicit ec: ExecutionContext)
+    extends FrontendController(mcc) with I18nSupport {
+
+  private val logger = Logger(this.getClass)
+
+  def displayQueryForm(): Action[AnyContent] = authenticate { implicit request =>
+    Ok(ileQueryPage(form))
+  }
+
+  def submitQueryForm(): Action[AnyContent] = authenticate { implicit request =>
+    form
+      .bindFromRequest()
+      .fold(
+        formWithErrors => BadRequest(ileQueryPage(formWithErrors)),
+        validUcr => Redirect(controllers.ileQuery.routes.IleQueryController.submitQuery(validUcr))
+      )
+  }
+
+  def submitQuery(ucr: String): Action[AnyContent] = authenticate.async { implicit request =>
+    ileQueryRepository.findBySessionIdAndUcr(retrieveSessionId, ucr).flatMap {
+      case Some(query) =>
+        connector.fetchQueryNotifications(query.conversationId, request.providerId).flatMap { response =>
+          response.status match {
+            case OK =>
+              val queryResponse = Json.parse(response.body).as[Seq[IleQueryResponseExchange]]
+
+              queryResponse match {
+                case Seq() => Future.successful(loadingPageResult)
+                case response +: _ =>
+                  ileQueryRepository.removeByConversationId(query.conversationId).map { _ =>
+                    displayQueryResults(response)
+                  }
+              }
+            case _ =>
+              logger.warn(s"Movements backend returned status: ${response.status}")
+              ileQueryRepository.removeByConversationId(query.conversationId).map { _ =>
+                InternalServerError(errorHandler.standardErrorTemplate())
+              }
+          }
+        }
+
+      case None => sendIleQuery(ucr)
+    }
+  }
+
+  private def loadingPageResult()(implicit request: Request[AnyContent]) =
```

Comment:
is this code is use somewhere else then submitQuery ? I think local def's are better then private methods if there single usage of this code.

Reply:
Yes, I agree with you, but now it's used in two different methods. Still see the pros of extracting this in one place to have the refresh value to change in one place.

I'll try to refactor this page after this PR